src/utils/http/fetchAPI.js

```diff
@@ -0,0 +1,56 @@
+import APIError from 'utils/http/errors/APIError';
+import getApiEndpoint from '../apiEndpoint';
+import getAuthJWT from '../getAuthJWT';
+import FetchError from './errors/FetchError';
+
+const fetchAPI = async (path, params = {}, extras = {}, returnValue = true) => {
+  const headers = params.headers ? params.headers : {};
+  const authJWT = extras.jwt || await getAuthJWT();
+
+  const parameters = {
+    ...params,
+    headers: {
+      ...headers,
+      ...(authJWT && { Authorization: `Bearer ${authJWT}` }),
+    },
+  };
+
+  const url = getApiEndpoint(extras.uiUrl) + path;
+
+  let response;
+  try {
+    response = await fetch(url, parameters);
+  } catch (e) {
+    // wrap fetch errors in custom error
+    throw new FetchError(e);
+  }
+
+  const isJson = response.headers.get('content-type')?.includes('application/json');
+
+  if (response.ok) {
+    // only parse response into JSON if needed because
+    // most patch / put reqs do not need a response
+    if (returnValue) {
+      return response.json();
+    }
+    return;
+  }
+  // console.log('response');
+  // console.log(response);
+  // if the response is not OK, parse json an return API error message
+  if (!response.ok) {
+    console.log('reponse type ', response.headers.get('content-type'));
+    console.log('response not ok, ', response, typeof (response));
+    const data = isJson ? await response.json() : null;
+    // data.message & data.errors follow error formatting defined in:
+    // HTTPError.v1.yaml
+    console.log('response not ok, raising APIError, ', data);
+    throw new APIError(response.status, data?.message, data?.errors);
+  }
+
+  // only parse response into JSON if needed because
+  // most patch / put reqs do not need a response
+  if (returnValue) return response.json();
```

Comment:
The previous fetchApi didn't perform the parsing, why is it better to parse inside fetchApi instead of leaving the parsing to the caller? Removing this .json() also means we can get rid of the returnValue parameter.

Also, I assume we are returning json always from fetchApi, but should we consider the possibility of us having to parse different types of files?

Reply:
it's better to do inside because **all** our calls do it so we avoid a lot of code. Moreover, I haven't done it here but it's possible to wrap those `json` calls in a `try/catch` to return a custom error like `JSONError` which then will also be handled by `handleError` abstracting away all the changes / handling of exceptions from the caller.

I don't think we should consider different types of files if we are not using them.

Reply:
I think perhaps the difference here is that I considered `fetchApi` as a wrapper for `fetch` that has some handling added, the final result when eveything works well is the same one as for `fetch`.

That's why having a `response.json()` inside kind of seemed weird to me, especially since that addition causes us to have to add a `returnValue` parameter to check whether to do it or not (because we are actually not doing `.json()` in all our calls after your refactoring, for example in `createCellSets`).

That's the main points I had, but I think if we keep it like this it should be ok as well. 

If we keep doing `response.json()` the only thing I would recommend perhaps is doing `return await reponse.json()` because of [this](https://stackoverflow.com/questions/44806135/why-no-return-await-vs-const-x-await/44806230#44806230).
Also disable the no-return-await rule from eslint, we already did that for the api.